core-bundle/src/Filesystem/Filesystem.php

```diff
@@ -0,0 +1,416 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of Contao.
+ *
+ * (c) Leo Feyer
+ *
+ * @license LGPL-3.0-or-later
+ */
+
+namespace Contao\CoreBundle\Filesystem;
+
+use Contao\StringUtil;
+use Contao\Validator;
+use League\Flysystem\Config;
+use League\Flysystem\DirectoryAttributes;
+use League\Flysystem\DirectoryListing;
+use League\Flysystem\FileAttributes;
+use League\Flysystem\FilesystemAdapter;
+use League\Flysystem\FilesystemOperator;
+use League\Flysystem\InvalidStreamProvided;
+use League\Flysystem\PathNormalizer;
+use League\Flysystem\StorageAttributes;
+use League\Flysystem\WhitespacePathNormalizer;
+use ProxyManager\Factory\LazyLoadingGhostFactory;
+use ProxyManager\Proxy\GhostObjectInterface;
+
+/**
+ * This Flysystem operator wraps a DBAFS layer around the underlying filesystem
+ * adaptor: Listing content will only access the database, operations that can
+ * mutate the filesystem, will automatically trigger a synchronization.
+ *
+ * Additionally, there is support to directly retrieve the 'extra metadata'
+ * from a file's / directory's StorageAttributes via a dedicated method.
+ *
+ * We widen the FilesystemOperator interface - instead of calling methods with
+ * resource paths (string), you can also pass in an ID (int) or UUID of a
+ * synchronized DBAFS resource.
+ *
+ * @phpstan-import-type ExtraMetadata from Dbafs
+ */
+final class Filesystem implements FilesystemOperator
+{
+    private FilesystemAdapter $adapter;
+    private Config $config;
+    private PathNormalizer $pathNormalizer;
+    private Dbafs $dbafs;
+    private LazyLoadingGhostFactory $proxyFactory;
+
+    public function __construct(Dbafs $dbafs, FilesystemAdapter $adapter, LazyLoadingGhostFactory $proxyFactory = null, Config $config = null, PathNormalizer $pathNormalizer = null)
+    {
+        $this->dbafs = $dbafs;
+        $this->adapter = $adapter;
```

Comment:
In order to really benefit from `Flysystem` I think we need to have the `Mount Manager` as first class citizen here. We don't have to implement it in core now but I think it's a fairly safe bet that we want to have this in the future. So that I can mount `/files/dropbox` to some dropbox account and `/files/s3` to some s3 bucket.
I'm not sure if the access details etc. have to go to the `Dbafs` or if it should reside in files like `.public` to ensure access is availalbe also without `Dbafs`. However, I'm rather sure, we cannot use the protocol specification `Flysystem` uses by default. Hence, I'd love to already address this now because now is the time to think about this conceptually.

Reply:
In the flysystem bundle there is a `\League\Flysystem\Filesystem` operator instance for each configured adapter + some magic in the extension, that makes the DI component inject the right one according to how you name your variable (so `default_storage` for `Filesystem $defaultStorage`). 

It's already possible if you manually configure it, but we could do the same magic for our operator for sure. You should then get another DBAFS service instance with its own database as well.

The mount manager is another layer on top of filesystem operators, so this should IMHO work out of the box (minus extra functionality built into ours like a flag to bypass the DBAFS).

Reply:
Yeah that's exactly what I'm saying. I don't think this makes sense. In our case, the only service any developer ever needs is the `VirtualFilesystem` and the different sources should be managed inside it based on the path. Hence we don't need any bundle imho.

Reply:
I started outlining a mount manager concept. I will update the description once there is more to see and then we can discuss based on this new approach.